src/app/clusters/operational-state-server/OperationalStateDataProvider.cpp

```diff
@@ -0,0 +1,264 @@
+                                                       size_t & size)
+{
+    uint8_t buffer[kOperationalStateMaxSerializedSize];
+    MutableByteSpan bufferSpan(buffer);
+    size           = 0;
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    ReturnErrorOnFailure(Load(DefaultStorageKeyAllocator::OperationalStatePhaseList(endpoint, clusterId).KeyName(), bufferSpan));
+
+    TLV::TLVReader reader;
+    TLV::TLVType outerType;
+
+    reader.Init(bufferSpan.data(), bufferSpan.size());
+    ReturnErrorOnFailure(reader.Next(TLV::TLVType::kTLVType_Array, TLV::AnonymousTag()));
+    ReturnErrorOnFailure(reader.EnterContainer(outerType));
+    size_t i                 = 0;
+    PhaseListCharSpan * head = nullptr;
+    while (reader.Next() != CHIP_ERROR_END_OF_TLV)
+    {
+        PhaseListCharSpan * newPhase = chip::Platform::New<PhaseListCharSpan>();
+        if (newPhase == nullptr)
+        {
+            ChipLogProgress(Zcl, "Malloc error");
+            break;
+        }
+        newPhase->Next = nullptr;
+
+        CharSpan readPhase;
+        ReturnErrorOnFailure(reader.Get(readPhase));
+
+        if (readPhase.size())
+        {
+            if (readPhase.size() <= kPhaseMaxSize)
+            {
+                char * dest = const_cast<char *>(newPhase->Phase);
+                size_t len  = readPhase.size();
+                memcpy(dest, readPhase.data(), len);
+                newPhase->phase = CharSpan::fromCharString(newPhase->Phase);
+            }
+            else
+            {
+                err = CHIP_ERROR_BUFFER_TOO_SMALL;
+            }
+        }
+        else
+        {
+            chip::Platform::Delete(newPhase);
+            continue;
+        }
+        i++;
+        if (head == nullptr)
+        {
+            head = newPhase;
+        }
+        else
+        {
+            PhaseListCharSpan * pList = head;
+            while (pList->Next != nullptr)
+            {
+                pList = pList->Next;
+            }
+            pList->Next = newPhase;
+        }
+    }
+
+    ReturnErrorOnFailure(reader.ExitContainer(outerType));
+    size       = i;
+    *phaseList = head;
+
+    return err;
+}
+
+void OperationalStateDataProvider::ReleasePhaseList(PhaseListCharSpan * phaseList)
+{
+    while (phaseList)
+    {
+        PhaseListCharSpan * del = phaseList;
+        phaseList               = phaseList->Next;
+        chip::Platform::Delete(del);
+    }
+}
+
+CHIP_ERROR OperationalStateDataProvider::ClearPhaseStateList(EndpointId endpoint, ClusterId clusterId)
+{
+    return mPersistentStorage->SyncDeleteKeyValue(
+        DefaultStorageKeyAllocator::OperationalStatePhaseList(endpoint, clusterId).KeyName());
+}
+
+CHIP_ERROR OperationalStateDataProvider::Load(const char * key, MutableByteSpan & buffer)
+{
+    uint16_t size = static_cast<uint16_t>(buffer.size());
+    ReturnErrorOnFailure(mPersistentStorage->SyncGetKeyValue(key, buffer.data(), size));
+
+    buffer = MutableByteSpan(buffer.data(), size);
```

Comment:
This is a weird way of writing `buffer.reduce_size(size);`

Reply:
Done.